system/src/asset_manager.cpp

```diff
@@ -0,0 +1,479 @@
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation, either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "hal_platform.h"
+
+#if HAL_PLATFORM_ASSETS
+
+#include "asset_manager.h"
+#include "storage_streams.h"
+#include "check.h"
+#include "ota_flash_hal_impl.h"
+#include "user_hal.h"
+#include <memory>
+#include "logging.h"
+#include "flash_mal.h"
+#include "endian_util.h"
+#include "scope_guard.h"
+#include "system_cache.h"
+
+namespace particle {
+
+namespace {
+
+const size_t FREE_BLOCKS_REQUIRED = 16;
+
+// TODO: move to using streams as well?
+int parseAssetDependencies(spark::Vector<Asset>& assets, hal_storage_id storageId, uintptr_t start, uintptr_t end) {
+    for (uintptr_t offset = start; offset < end;) {
+        module_info_extension_t ext = {};
+        hal_storage_read(storageId, offset, (uint8_t*)&ext, sizeof(ext));
+        if (ext.type == MODULE_INFO_EXTENSION_ASSET_DEPENDENCY) {
+            auto assetExt = std::unique_ptr<module_info_asset_dependency_ext_t>((module_info_asset_dependency_ext_t*)calloc(1, ext.length + 1 /* NULL name terminator */));
+            CHECK_TRUE(assetExt, SYSTEM_ERROR_NO_MEMORY);
+            CHECK(hal_storage_read(storageId, offset, (uint8_t*)assetExt.get(), ext.length));
+            if (assetExt->hash.type != MODULE_INFO_HASH_TYPE_SHA256) {
+                LOG(WARN, "Unsupported hash type %u with length %u", assetExt->hash.type, assetExt->hash.length);
+                continue;
+            }
+            if (assetExt->hash.type != AssetHash::DEFAULT) {
+                LOG(WARN, "Unknown hash type for asset dependency: 0x%x", assetExt->hash.type);
+            }
+            assets.append(Asset(assetExt->name, AssetHash(assetExt->hash.hash, assetExt->hash.length)));
+        } else if (ext.type == MODULE_INFO_EXTENSION_END) {
+            break;
+        }
+        offset += ext.length;
+    }
+    return 0;
+}
+
+int parseAssetInfo(InputStream* stream, size_t size, Asset& asset) {
+    std::unique_ptr<module_info_name_ext_t> nameExt;
+    AssetHash hash;
+    while (size > 0) {
+        module_info_extension_t ext = {};
+        CHECK(stream->peek((char*)&ext, sizeof(ext)));
+        CHECK_TRUE(ext.length <= size, SYSTEM_ERROR_BAD_DATA);
+        if (ext.type == MODULE_INFO_EXTENSION_NAME) {
+            CHECK_TRUE(ext.length > sizeof(module_info_name_ext_t), SYSTEM_ERROR_BAD_DATA);
```

Comment:
Just verifying `>` here and `>=` (line 78) are intentionally different.

Reply:
Using `>` would allow for a minimum of one additional character, which is probably null.